### PR TITLE
`Learning paths`: Refactor competency graph to get data passed from outside of component

### PR DESCRIPTION
**Pull request by JohannesWt**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.

### Motivation and Context

The competency graph can only be used to show the mastery progress at the moment. This should be generalised to show also different data.

### Description

The competency graph now gets the data passed from outside so that different data can be shown using the same graph. Additionally, the returned data of the server has been refactored to return a value with a corresponding type. This way, the competency graph can also be used to show aggregated data in the learning paths instructor analytics section. This PR changed nothing visually.

### Steps for Testing

Prerequisites:
- 1 student
- 1 course with competencies 
- at least 1 lecture with at least 1 lecture unit (or a Programming Exercise)

1. Log in to Artemis
2. Navigate to Course 
3. Navigate to learning path student UI
4. open the learning objects overview dropdown
5. open the competency graph
6. verify that the competency graph is still shown with all competencies of the course

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| competency-graph-modal.component.ts | 100% | ✅ |
| competency-graph.component.ts | 79.16% | ✅  |
| competency-node.component.ts | 100% | ✅ |
| learning-path.model.ts | 96.42% | ✅  |

### Screenshots

UI is unchanged

## Summary by CodeRabbit

- **New Features**
  - Enhanced loading state in the competency graph modal for a better user experience.
  - Simplified competency representation with new value types for clearer data handling.

- **Bug Fixes**
  - Improved clarity and focus in competency progress assertions in tests.

- **Refactor**
  - Streamlined component logic and data handling in the competency graph and node components.
  - Updated competency node display logic for better clarity on competency states.
  - Improved calculation and representation of mastery progress in the competency graph.

- **Tests**
  - Introduced a comprehensive test suite for the competency graph modal, enhancing test coverage.
  - Adjusted tests in the competency graph and node components to align with new data structures.

### CODE REVIEW

**Review by Jan-Thurner — APPROVED** [review]
tested on TS2. Works as expected.

**Review by az108 — CHANGES_REQUESTED** [review]
I got following while testing.
image

**Review by pzdr7 — COMMENTED** [review]
The `CompetencyNodeValueType` enum currently has one value - how soon are we going to extend it to have more values?

File: src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathService.java [hunk 1]
```diff
@@ -307,8 +307,11 @@ public LearningPathCompetencyGraphDTO generateLearningPathCompetencyGraph(@NotNu
         Map<Long, CompetencyProgress> competencyProgresses = competencyProgressRepository.findAllByCompetencyIdsAndUserId(competencyIds, user.getId()).stream()
                 .collect(Collectors.toMap(progress -> progress.getCompetency().getId(), cp -> cp));
 
-        Set<CompetencyGraphNodeDTO> progressDTOs = competencies.stream()
-                .map(competency -> CompetencyGraphNodeDTO.of(competency, Optional.ofNullable(competencyProgresses.get(competency.getId())))).collect(Collectors.toSet());
+        Set<CompetencyGraphNodeDTO> progressDTOs = competencies.stream().map(competency -> {
+            var competencyProgressOptional = Optional.ofNullable(competencyProgresses.get(competency.getId()));
+            var masteryProgress = competencyProgressOptional.isPresent() ? CompetencyProgressService.getMasteryProgress(competencyProgresses.get(competency.getId())) : 0;
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathService.java:312]:
```suggestion
            var masteryProgress = competencyProgressOptional.map(CompetencyProgressService::getMasteryProgress).orElse(0.0);
```

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
Code looks good overall, left one small improvement:

**Review by pzdr7 — APPROVED** [review]
Thanks for implementing the changes 👍

**Review by Jan-Thurner — APPROVED** [review]
tested again on TS1. Still works for me.

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by az108 — APPROVED** [review]
Tested on TS1 works as expected now 👍 
image

**Review by MaximilianAnzinger — APPROVED** [review]
Maintainer approved

**Pull request merged**